### PR TITLE
add loading indicators

### LINKED ISSUE #519: Add loading indicator on export tab

### LINKED ISSUE #576: Show loading indicator when initializing charts
Per suggestion in #538 from @wal615. Related to #519 as well.

### PR DESCRIPTION
**Pull request by jwildfire**
# Overview 

Added 2 types of loading indicators using `shinybusy`: 
- Added a basic loading indicator that runs whenever server runs for more than 1 second (like when rendering a big chart)
- Added a modal indicator that pops up while html reports are being generated. This prevents other activity in the app, which wouldn't work anyway since the server is busy rendering the report. 

Fix #576. Fix #519.

# Test notes

Make sure the new loading indicators are working as expected and don't have any major unintended consequences ...

### CODE REVIEW

File: R/mod_safetyGraphicsUI.R [hunk 1]
```diff
@@ -32,6 +33,7 @@ safetyGraphicsUI <- function(id, meta, domainData, mapping, standards){
     #app UI using calls to modules
     ui<-tagList(
         shinyjs::useShinyjs(),
+        add_busy_spinner(spin = "atom", position="bottom-right"),
```
Comment (samussiah) [inline R/mod_safetyGraphicsUI.R:36]: That spinner sure is pretty.

File: R/mod_reportsTab.R [hunk 2]
```diff
@@ -84,12 +85,13 @@ reportsTab <- function(input, output, session, charts, data, mapping){
...
         charts=charts_keep()
       )
-      
+      show_modal_spinner(text="Creating html report. This might take a while for multiple charts/large data sets.") # show the modal window
       rmarkdown::render(tempReport,
                         output_file = file,
                         params = params,  ## pass in params
                         envir = new.env(parent = globalenv())  ## eval in child of global env
       )
+      remove_modal_spinner() # remove it when done
```
Comment (samussiah) [inline R/mod_reportsTab.R:94]: This line doesn't run if report fails - Demographics Table, Shift Table - and spinner just spins.

image

Comment (samussiah) [reply R/mod_reportsTab.R:94]: Can we just drop these two modules for this release?  They don't work on my workstation.

**Review by samussiah — CHANGES_REQUESTED** [review]
Spinner locks the app when export fails.

File: R/mod_reportsTab.R (hunk 2 above)
Comment (jwildfire) [reply R/mod_reportsTab.R:94]: Yeah, Shift Table is no longer in the default build after https://github.com/SafetyGraphics/safetyCharts/pull/67

Comment (jwildfire) [reply R/mod_reportsTab.R:94]: > This line doesn't run if report fails - Demographics Table, Shift Table - and spinner just spins.

Yeah, I'm not sure the best workflow here. I guess at the very least we should have some trycatch() logic that throws a decent message to the modal if the export fails.

I think I'm going to leave #517 for a different PR that deals with some of the other issues in the export. I'll close this one, and make a new PR that just resolves #576.

**Pull request closed without merging**